nvidia_gpu_stats/src/main.rs

```diff
@@ -0,0 +1,356 @@
+            metrics.insert(format!("_gpu.{}.pcieLinkSpeed", di), json!(link_speed));
+        }
+
+        let link_width = device.current_pcie_link_width()?;
+        metrics.insert(format!("_gpu.{}.pcieLinkWidth", di), json!(link_width));
+
+        let max_link_gen = device.max_pcie_link_gen()?;
+        metrics.insert(format!("_gpu.{}.maxPcieLinkGen", di), json!(max_link_gen));
+
+        let max_link_width = device.max_pcie_link_width()?;
+        metrics.insert(
+            format!("_gpu.{}.maxPcieLinkWidth", di),
+            json!(max_link_width),
+        );
+
+        let cuda_cores = device.num_cores()?;
+        metrics.insert(format!("_gpu.{}.cudaCores", di), json!(cuda_cores));
+
+        let architecture = device.architecture()?;
+        metrics.insert(
+            format!("_gpu.{}.architecture", di),
+            json!(format!("{:?}", architecture)),
+        );
+    }
+
+    Ok(GpuMetrics { metrics })
+}
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // Parse command-line arguments
+    let args = Args::parse();
+
+    // Initialize NVML
+    let nvml_result = nvml_wrapper::Nvml::init();
+
+    // Set up a flag to control the main sampling loop
+    let running = Arc::new(AtomicBool::new(true));
+    let r = running.clone();
+
+    // Set up signal handler for graceful shutdown
+    let mut signals = Signals::new(TERM_SIGNALS)?;
+    thread::spawn(move || {
+        for _sig in signals.forever() {
+            r.store(false, Ordering::Relaxed);
+            break;
+        }
+    });
+
+    // Main sampling loop. Will run until the parent process is no longer alive or a signal is received.
+    while running.load(Ordering::Relaxed) {
+        let sampling_start = Instant::now();
+        let timestamp = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_secs_f64();
+
+        // Sample GPU metrics
+        let mut gpu_metrics = match &nvml_result {
+            Ok(nvml) => match nvml.sys_cuda_driver_version() {
+                Ok(cuda_version) => {
+                    let cuda_version = format!(
+                        "{}.{}",
+                        nvml_wrapper::cuda_driver_version_major(cuda_version),
+                        nvml_wrapper::cuda_driver_version_minor(cuda_version)
+                    );
+                    match sample_metrics(nvml, args.pid, cuda_version) {
+                        Ok(metrics) => metrics,
+                        Err(_) => sample_metrics_fallback(),
+                    }
+                }
+                Err(_) => sample_metrics_fallback(),
+            },
+            Err(_) => sample_metrics_fallback(),
+        };
+
+        // Add timestamp to metrics
+        gpu_metrics
+            .metrics
+            .insert("_timestamp".to_string(), json!(timestamp));
+
+        // Convert metrics to JSON and print to stdout for collection
+        let json_output = serde_json::to_string(&gpu_metrics.metrics).unwrap();
```

Comment:
This can panic as well, but again it might be OK here.

It would be great if we could monitor these errors.

Reply:
left a todo for now. if this effort is successful, I want to move all system metrics into this external package and make it a proper service, with proper logging.